examples/bone_finetuning/finetune.py

```diff
@@ -0,0 +1,385 @@
+            model = PeftModel.from_pretrained(model, script_args.model_name_or_path, subfolder = script_args.adapter_name_or_path, is_trainable=True, config=bone_config)
+        else:
+            logger.info(f'Init LoRA modules...')
+        peft_config = BoneConfig(
+            task_type=TaskType.CAUSAL_LM,
+            target_modules=script_args.target_modules.split(","),
+            inference_mode=False,
+            r=script_args.bone_r,
+            init_weights=script_args.init_bone_weights,
+        )
+        model = get_peft_model(model, peft_config, adapter_name="weight")
+
+    for name, module in model.named_modules():
+        if "norm" in name or "gate" in name:
+            module = module.to(torch.float32)
+    return model
+
+
+def train():
+    parser = transformers.HfArgumentParser(TrainingArguments)
+    script_args = parser.parse_args_into_dataclasses()[0]
+    log_level = script_args.get_process_log_level()
+    logger.setLevel(log_level)
+    datasets.utils.logging.set_verbosity(log_level)
+    transformers.utils.logging.set_verbosity(log_level)
+    transformers.utils.logging.enable_default_handler()
+    transformers.utils.logging.enable_explicit_format()
+
+    if script_args.local_rank == 0:
+        logger.info("=" * 100)
+        logger.info(script_args)
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        script_args.model_name_or_path,
+        model_max_length=script_args.model_max_length,
+        padding_side="right",
+        use_fast=True,
+        trust_remote_code=True,
+    )
+    if tokenizer.pad_token is None:
+        tokenizer.pad_token = tokenizer.eos_token
+    logger.info("PAD Token:", tokenizer.pad_token, tokenizer.pad_token_id)
+    logger.info("BOS Token", tokenizer.bos_token, tokenizer.bos_token_id)
+    logger.info("EOS Token", tokenizer.eos_token, tokenizer.eos_token_id)
+
+    if script_args.local_rank == 0:
+        logger.info(f"Load tokenizer from {script_args.model_name_or_path} over.")
+
+    resume_from_checkpoint_dir = get_last_checkpoint(script_args.output_dir)
+    model = build_model(script_args, resume_from_checkpoint_dir)
+
+    raw_train_datasets = load_dataset(script_args.data_path, split=script_args.dataset_split)
+
+    if script_args.local_rank > 0:
+        torch.distributed.barrier()
+
+    train_dataset = raw_train_datasets.map(
+        train_tokenize_function,
+        batched=True,
+        batch_size=3000,
+        num_proc=32,
+        remove_columns=raw_train_datasets.column_names,
+        load_from_cache_file=True,
+        desc="Running tokenizer on train dataset",
+        fn_kwargs={
+            "tokenizer": tokenizer,
+            "query": script_args.dataset_field[0],
+            "response": script_args.dataset_field[1],
+        },
+    )
+
+    if script_args.local_rank == 0:
+        torch.distributed.barrier()
```

Comment:
I tried running this script locally without DeepSpeed and this line raised an error because "Default process group has not been initialized". I wonder if it can be omitted or worked around so that users can easily call this script without distributed training.

Reply:
I have provided comments for the training script without DeepSpeed in the scripts (due to my setup limitations, I am unable to test it accurately).